src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs

```diff
@@ -888,5 +952,40 @@ private static void ThrowNotAuthenticated()
         {
             throw new InvalidOperationException(SR.net_auth_noauth);
         }
+
+        // (non-generic) copy of the PointerMemoryManager<T> which supports resetting the stored
+        // pointer to allow pooling its instances instead of allocating a new one per Read/Write call.
+        // The memory ponted to by the intenal poiner is assumed to be externally pinned (or naive memory).
+        internal sealed unsafe class PoolingPointerMemoryManager : MemoryManager<byte>
+        {
+            private void* _pointer;
+            private int _length;
+
+            protected override void Dispose(bool disposing)
+            {
+            }
+
+            public void Reset(void* pointer, int length)
+            {
+                _pointer = pointer;
+                _length = length;
+            }
+
+            public override Span<byte> GetSpan()
+            {
+                return new Span<byte>(_pointer, _length);
+            }
+
+            public override MemoryHandle Pin(int elementIndex = 0)
+            {
+                // memory assumed to be pinned already
+                return new MemoryHandle(_pointer, default, null);
```

Comment:
This isn't factoring in elementIndex.

Reply:
Thanks, I completely overlooked this.

Interestingly enough, I was unable to reproduce the failures locally (on the same OS).